docs/rfcs/generic-record-sync.md

```diff
@@ -0,0 +1,906 @@
+    ]
+}
+```
+
+See the comments in the rust code rust code for a description of what this would
+mean, what restrictions we'd impose, etc.
+
+## Guid Renames
+
+In order to support references to other IDs, we primarially need to handle
+deduping. E.g, the guid on the server may be what a record used to be called,
+however now it has a differnt name. My recommendation is a couple small changes
+to the server protocol to support it, however we could also implement this
+ourselves with more difficulty (and lack of transactional semantics).
+
+### Server support
+
+This is two parts:
+
+1. Add a single optional field to BSOs: `prev_id`.
+    - This is an optional field indicates that `prev_id` has been renamed to
+      `id`, and that (once the batch is committed).
+    - `prev_id` does not need to reference an ID that is known to the server,
+      however it makes sense to ensure it's a valid BSO ID string.
+    - This would not change the semantics of any existing APIs. Critically,
+      requests to the old ID should still return whatever is (or is not) present
+      for that id.
+
+2. New endpoint: `GET https://<endpoint-url>/rename?ids=<ids>`
+    - `ids` is a comma separated list of up to 100 (as with
+      `{DELETE,GET} /storage/collection?ids=`)
+    - Returns a JSON array of the renamed version of the ids, in the same order
+      they were sent.
+    - For IDs that were never renamed, or that are not known, should return them
+      as-is, and for IDs that have been renamed, it should return the new ID.
+      This is transitive.
+    - Notes:
+        - Renames are not collection specific. Any collection is allowed to
+          store an ID from any other collection!
+        - The length of the array should always be identical to the number of
+          ids requested.
+        - Renames should be resolved transitively, e.g. if `"a"` is renamed to
+          `"b"`, and `"b"` is renamed to `"c"`, then `/rename?ids=a,b` should
+          return `["c", "c"]`.
+        - Presumably the table this is stored in on the server would have both
+          `(<user_id>, a, c)` and `(<user_id>, b, c)` in it, or something along
+          those lines.
+    - Some thought is needed about how to expire this data. Initially (I have
+      not thought for very long about this):
+        - It seems reasonable to me to delete the data about the renames when
+          the *transitive* destination ID is deleted (e.g. the case above should
+          still work even if "b" has been deleted, but once "c" is deleted it
+          seems like it can be deleted).
+
+### Legacy client support
+
+We would need to extend legacy clients to provide `prev_id` in the case of
+renames. This seems reasonably easy to do on a best-effort basis (store the data
+in memory, or even back with JSONFile).
+
+## Clocks, Counters, Etc.
+
+For most of this I'm assuming we're using lamport clocks, e.g. a single counter
```

Comment:
Actually, I think I',m going to change things so that we use vector clocks. They aren't that complex, and they will let us avoid falsely assuming conflicts occur when they have not. They also allow us to perform two-way-merges with a lot more confidence.